### ISSUE TITLE
Grad Caching For MNR Loss

### ISSUE THREAD

**Issue by johneckberg** [UKPLab/sentence-transformers#2428]
As mentioned in the documentation, MNR loss benefits from larger batch sizes as it allows for more negative samples. However, this presents an issue in a memory constrained setting as these larger batch sizes can't fit in the gpu memory and standard gradient accumulation doesn't work. The paper [Scaling Deep Contrastive Learning Batch Size under Memory Limited Setup](https://arxiv.org/abs/2101.06983) provides a nice solution to this issue, outlining a gradient accumulation algorithm for MNR loss. Is this a feature that could be added to MNR Loss? I would be glad to tackle this and put in a PR.

**Comment by tomaarsen** [comment]
Hello!

A few days ago, I merged #1759 by @kwang2049 that implemented the CachedMultipleNegativeRankingLoss: https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers%2Flosses%2FCachedMultipleNegativesRankingLoss.py

It'll be included in the upcoming release, and as it's not released yet, it also hasn't been added to the website quite yet.

That said, you're free to review the implementation for potential inefficiencies! And alternatively, if you're in the mood for implementing a loss function, you can look into the [AnglE loss](https://github.com/SeanLee97/AnglE). I haven't had time to work on the implementation in ST yet.

- Tom Aarsen

**Comment by ir2718** [comment]
@tomaarsen 

Hi Tom, could you publish a list of stuff you would like to implement in ST in the long run so people could have a go at contributing?

**Comment by johneckberg** [comment]
@tomaarsen 

I would be happy to implement AnglE loss!  In the paper, the objective used is combination of what the authors call [Cosine loss]( https://github.com/bojone/CoSENT/blob/124c368efc8a4b179469be99cb6e62e1f2949d39/cosent.py#L79), standard MNR loss, and Angle loss. Should I implement both Cosine and Angle loss, or just Angle loss?

Also, seconding @ir2718, a list of implementation to-dos would be fantastic!

**Comment by tomaarsen** [comment]
That's awesome! I haven't delved into the details of the cosine loss myself, but I assumed it was equivalent to the CosineSimilarityLoss from ST. If that's indeed the case, then we only need the AnglE loss.
Afterwards, we can try and determine if there is a convenient way for users to specify multiple (weighted) losses at once, much like the AnglE paper proposes.

As for the todo-list for the project - that's a great idea. I'll come up with a more detailed list during the next week. I'll @ both of you with the results! 

- Tom Aarsen

**Comment by johneckberg** [comment]
Sounds great Tom! I was able to take a closer look at the paper and the code, and can confirm what is called cosine loss is different than CosineSimilarityLoss from ST. It traces back to this [blog post](https://kexue.fm/archives/8847), where it is described by the author as a "sorting loss".

![image](https://github.com/UKPLab/sentence-transformers/assets/93222362/247f4363-9d32-45a8-9dcb-0b7adf5fd9b8)

 I'll implement both losses for ST, and make note of any other unusualities in the PR!
